### PR TITLE
Initial implementation of np.linalg.lstsq() via SVD

### PR DESCRIPTION
**Pull request by jakevdp**
This is an initial implementation of ``np.linalg.lstsq`` based on the SVD. A full solution would involve adding wrappers for ``*gelsd`` to lax_linalg.py & cusolver.

I estimate this is about 2x slower than the full solution, based on performance of the relevant lapack code paths in numpy.

Addresses part of #1999

### CODE REVIEW

File: jax/numpy/linalg.py [hunk 1]
```diff
@@ -367,3 +367,46 @@ def solve(a, b):
...
+  if rcond is None:
+    rcond = np.finfo(dtype).eps * max(n, m)
+  elif rcond < 0:
+    rcond = np.finfo(dtype).eps
+  u, s, vt = svd(a, full_matrices=False)
+  mask = s >= rcond * s[0]
+  rank = mask.sum()
+  s_inv = np.where(mask, 1 / s, 0)
+  x = vt.conj().T @ (u.conj().T @ b * s_inv[:, None])
+  resid = norm(b - a @ x, axis=0) ** 2
```
Comment (shoyer) [inline jax/numpy/linalg.py:405]: Until we change the default precision of `matmul` (see https://github.com/google/jax/issues/2161), it's probably best to stick to explicitly calling `jnp.matmul(..., precision=lax.Precision.HIGHEST`. Otherwise these linear algebra routines tend to have absolutely horrific accuracy on TPUs.

File: jax/numpy/linalg.py [hunk 2]
```diff
@@ -367,3 +367,46 @@ def solve(a, b):
...
+  # To match numpy, we would add the following condition. We don't do
+  # this because it makes lstsq incompatible with jit.
+  if numpy_resid and (rank < n or m <= n):
+      resid = np.asarray([])
```
Comment (shoyer) [inline jax/numpy/linalg.py:409]: Presumably we should put the other calculation of `resid` in an `else` block, to avoid calculating it if it isn't used?

File: jax/numpy/linalg.py [hunk 3]
```diff
@@ -367,3 +367,46 @@ def solve(a, b):
...
+  b_ndim = b.ndim
+  if b_ndim == 1:
+    b = b[:, None]
+  if a.ndim != 2:
+    raise TypeError(
+      f"{a.ndim}-dimensional array given. Array must be two-dimensional")
+  if b.ndim != 2:
+    raise TypeError(
+      f"{b_ndim}-dimensional array given. Array must be one or two-dimensional")
+  m, n = a.shape
```
Comment (shoyer) [inline jax/numpy/linalg.py:395]: There should probably also be a check verifying that `a.shape[-2] == b.shape[-2]`

File: jax/numpy/linalg.py [hunk 4]
```diff
@@ -367,3 +367,46 @@ def solve(a, b):
...
+  if rcond is None:
+    rcond = np.finfo(dtype).eps * max(n, m)
+  elif rcond < 0:
+    rcond = np.finfo(dtype).eps
+  u, s, vt = svd(a, full_matrices=False)
+  mask = s >= rcond * s[0]
+  rank = mask.sum()
+  s_inv = np.where(mask, 1 / s, 0)
```
Comment (shoyer) [inline jax/numpy/linalg.py:403]: I think you might need to use the "double where" trick to ensure that the gradient is well defined for singular values of zero, i.e.,
```suggestion
  safe_s = np.where(mask, s, 1)
  s_inv = np.where(mask, 1 / safe_s, 0)
```

File: jax/numpy/linalg.py [hunk 5]
```diff
@@ -367,3 +367,46 @@ def solve(a, b):
...
+       with jit. The non-jit compatible numpy behavior can be recovered by passing
+       numpy_resid=True.
+    """))
+def lstsq(a, b, rcond=None, *, numpy_resid=False):
+  # TODO: add lstsq to lax_linalg and implement this function via those wrappers.
+  a, b = _promote_arg_dtypes(a, b)
+  dtype = a.dtype
+  b_ndim = b.ndim
+  if b_ndim == 1:
+    b = b[:, None]
```
Comment (shoyer) [inline jax/numpy/linalg.py:388]: I go back and forth on this, but arguably `newaxis` is more readable:
```suggestion
    b = b[:, np.newaxis]
```

File: tests/linalg_test.py [hunk 6]
```diff
@@ -783,6 +783,37 @@ def testMultiDot(self, shapes, dtype, rng_factory):
...
+    onp_fun = partial(onp.linalg.lstsq, rcond=rcond)
+    jnp_fun = partial(np.linalg.lstsq, rcond=rcond)
+    jnp_fun_numpy = partial(np.linalg.lstsq, rcond=rcond, numpy_resid=True)
+    args_maker = lambda: [rng(lhs_shape, dtype), rng(rhs_shape, dtype)]
+    tol = {onp.float32: 1e-6, onp.float64: 1e-12,
+           onp.complex64: 1e-6, onp.complex128: 1e-12}
+
+    self._CheckAgainstNumpy(onp_fun, jnp_fun_numpy, args_maker, check_dtypes=False, tol=tol)
+    self._CompileAndCheck(jnp_fun, args_maker, check_dtypes=True, atol=tol, rtol=tol)
+
```
Comment (shoyer) [inline tests/linalg_test.py:816]: I think it would be a good idea to add a `check_grads()` call here to verify gradients, given how fragile gradients are for some of the operations involved here (like SVD or potential division by zero)

File: tests/linalg_test.py [hunk 7]
```diff
@@ -783,6 +783,37 @@ def testMultiDot(self, shapes, dtype, rng_factory):
...
+           jtu.format_shape_dtype_string(lhs_shape, dtype),
+           jtu.format_shape_dtype_string(rhs_shape, dtype),
+           rcond),
+       "lhs_shape": lhs_shape, "rhs_shape": rhs_shape, "dtype": dtype,
+       "rcond": rcond, "rng_factory": rng_factory}
+      for lhs_shape, rhs_shape in [
+          ((1, 1), (1, 1)),
+          ((4, 6), (4,)),
+          ((6, 6), (6, 1)),
+          ((8, 6), (8, 4)),
```
Comment (shoyer) [inline tests/linalg_test.py:798]: I think it would also be a good idea to test co-linear and/or duplicate columns in the design matrix `a`, which you won't get with random numbers. I think these would correspond to repeated singular values in the SVD? (repeated singular values will probably break the SVD JVP rule)

File: jax/numpy/linalg.py (hunk 1 above)
Comment (jakevdp) [reply jax/numpy/linalg.py]: Done

File: jax/numpy/linalg.py (hunk 3 above)
Comment (jakevdp) [reply jax/numpy/linalg.py:396]: Done

File: jax/numpy/linalg.py (hunk 4 above)
Comment (jakevdp) [reply jax/numpy/linalg.py]: Done

File: tests/linalg_test.py (hunk 6 above)
Comment (jakevdp) [reply tests/linalg_test.py:822]: I gave this a try... it fails with a dtype casting error in test_util. Perhaps due to the fact that some of the outputs are integer type?

I commented-out the test for now.

File: tests/linalg_test.py (hunk 7 above)
Comment (jakevdp) [reply tests/linalg_test.py:798]: Done

File: jax/numpy/linalg.py (hunk 5 above)
Comment (jakevdp) [reply jax/numpy/linalg.py:389]: Done

File: jax/numpy/linalg.py (hunk 2 above)
Comment (jakevdp) [reply jax/numpy/linalg.py]: Done

File: tests/linalg_test.py (hunk 6 above)
Comment (shoyer) [reply tests/linalg_test.py:822]: Yes, that makes sense. I would suggest just testing the first output from `lstsq` then, via a wrapper function.

Comment (shoyer) [reply tests/linalg_test.py:822]: e.g., `jtu.check_grads(lambda *args: jnp_fun(*args)[0], args_maker(), order=2, atol=tol, rtol=tol)`

Comment (jakevdp) [reply tests/linalg_test.py:821]: Thanks! That works - 32-bit was having issues, so I limited it to 64-bit similar to the check in Cholesky.

**Comment by jakevdp** [comment]
CI failure looks like a conda failure. Can someone restart it?

**Comment by jakevdp** [comment]
nvm, once I logged-in to Travis I was able to restart.

**Comment by joaogui1** [comment]
Hey, what about #2200 ?

**Comment by mattjj** [comment]
Uh oh, I forgot about #2200 :/ This is completely my bad.

I'm really sorry for the repeated work, and for accidentally ignoring the work you put in @joaogui1 . As JAX activity has picked up (especially inside Alphabet) we've gotten a lot worse at following up on PRs from amazing OSS contributors. (OSS contributions are especially amazing because contributors don't have the benefit of our extremely-active internal chat channels.)

We're trying to address the general problem. As of this week we're experimenting with a GitHub rotation. It's tough, though, because the JAX team is pretty small. I'm optimistic that we'll get better over time.

As for this specific case: @jakevdp @joaogui1 is there a way to combine efforts here and maybe draw on both PRs? Or is there too much redundancy and we need to chalk this up to a mistake to learn from?

@joaogui1 let me know if there is some course of action I can take to make this more right. (Also, maybe this is a good chance to highlight any other PRs of yours that we've let languish...)

**Comment by joaogui1** [comment]
So, after reading his code I believe @jakevdp implementation is better than mine, so I will close my PR.   
I have two other open PRs:

-  #1496 , frankly there are implementations of it in both haiku and flax, so I think I will also close this one. If I may say I think there should be a section in Contributing.md explaining that stax is not a priority for PRs (if it's accepting them at all) as it is more a proof of concept. I mean [there](https://github.com/google/jax/pull/2157) [are](https://github.com/google/jax/pull/2153)[some](https://github.com/google/jax/pull/2090) [dead](https://github.com/google/jax/pull/1569) PRs related to it, and Jake Bradbury said: 

> ... stax.py is meant more as an example (forking encouraged!) than as a growing or comprehensive library ...

- The other one is #1862 which I still think is important, but I would need more input from the dev team, so I understand if it will take a while longer for someone to pop up there and help me with deciding what tests are needed

Also, if you guys want some help I can reopen #1874 and search for solved/obsolete issues and PRs so someone can close them and help organize things a little more

**Comment by jakevdp** [comment]
I'm not able to get check_grads to pass consistently. Tt seems to be flaking on complext128 input, and/or low-rank inputs – 7 failures under `--num_generated_cases=25`. The gradient computation and its test is still a bit of a black box to me, and I haven't made much progress in debugging the issue.

**Comment by hawkinsp** [comment]
Is it worth merging this even without gradients?

(It seems that the problem is most likely not with the PR itself, given it doesn't implement any new gradients.)

**Comment by jakevdp** [comment]
> Is it worth merging this even without gradients?

Perhaps – I'm planning on digging-in to implement gradients via a custom jvp, but I won't have time to look closely at that until next week.

**Comment by jakevdp** [comment]
In offline discussion with @mattjj, we decided it would be worth submitting this even without robust gradient support. I removed the gradient test for now, and marked it TODO.

**Comment by shoyer** [comment]
Sounds good!

**Comment by jakevdp** [comment]
I just rebased on master to pick up changes there.

PTAL - I think this is ready for a final review.

**Review by shoyer — APPROVED** [review]
